services/autoscaling/src/simcore_service_autoscaling/utils_aws.py

```diff
@@ -2,105 +2,221 @@
+    client = None
+    try:
+        client = boto3.client(
+            "ec2",
+            endpoint_url=settings.EC2_ENDPOINT,
+            aws_access_key_id=settings.EC2_ACCESS_KEY_ID,
+            aws_secret_access_key=settings.EC2_SECRET_ACCESS_KEY,
+            region_name=settings.EC2_REGION_NAME,
+        )
+        yield client
+    finally:
+        if client:
+            client.close()
+
+
+def get_ec2_instance_capabilities(
+    settings: EC2Settings, instance_settings: EC2InstancesSettings
+) -> list[EC2Instance]:
+    with ec2_client(settings) as ec2:
+        instance_types = ec2.describe_instance_types(
+            InstanceTypes=instance_settings.EC2_INSTANCES_ALLOWED_TYPES
+        )
+
+    list_instances: list[EC2Instance] = []
+    for instance in instance_types.get("InstanceTypes", []):
+        with contextlib.suppress(KeyError):
+            list_instances.append(
+                EC2Instance(
+                    name=instance["InstanceType"],
+                    cpus=instance["VCpuInfo"]["DefaultVCpus"],
+                    ram=parse_obj_as(
+                        ByteSize, f"{instance['MemoryInfo']['SizeInMiB']}MiB"
+                    ),
+                )
+            )
+    return list_instances
+
+
+def closest_instance_policy(
+    ec2_instance: EC2Instance,
+    resources: Resources,
+) -> float:
+    if ec2_instance.cpus < resources.cpus or ec2_instance.ram < resources.ram:
+        return 0
+    # compute a score for all the instances that are above expectations
+    # best is the exact ec2 instance
+    cpu_ratio = float(ec2_instance.cpus - resources.cpus) / float(ec2_instance.cpus)
+    ram_ratio = float(ec2_instance.ram - resources.ram) / float(ec2_instance.ram)
+    return 100 * (1.0 - cpu_ratio) * (1.0 - ram_ratio)
+
+
+def find_best_fitting_ec2_instance(
+    available_ec2_instances: list[EC2Instance],
+    resources: Resources,
+    score_type=closest_instance_policy,
+) -> EC2Instance:
+    score_to_ec2_candidate: dict[float, EC2Instance] = OrderedDict(
+        sorted(
+            {
+                score_type(instance, resources): instance
+                for instance in available_ec2_instances
+            }.items(),
+            reverse=True,
+        )
     )
+    if not score_to_ec2_candidate:
+        raise Ec2InstanceNotFoundError(needed_resources=resources)
+    score, instance = next(iter(score_to_ec2_candidate.items()))
+    if score == 0:
+        raise Ec2InstanceNotFoundError(
+            needed_resources=resources, msg="no adequate EC2 instance found!"
+        )
+    return instance
 
 
-def create_ec2_client(settings: AwsSettings):
-    return boto3.client(
-        "ec2",
-        aws_access_key_id=settings.AWS_ACCESS_KEY_ID,
-        aws_secret_access_key=settings.AWS_SECRET_ACCESS_KEY,
-        region_name=settings.AWS_REGION_NAME,
+def _compose_user_data(docker_join_script: str) -> str:
+    return dedent(
+        f"""\
+#!/bin/bash
+{docker_join_script}
+"""
```

Comment:
I think this works with decent if I am not wrong
```suggestion
    return dedent(
        f"""\
        #!/bin/bash
        {docker_join_script}
        """
```

Reply:
ok, so this does not work. I do have a test to prove it if you wanna try.